### ISSUE TITLE
Making the case for MemoryMarshal.Cast

### ISSUE THREAD

**Issue by rileyzzz** [Facepunch/sbox-issues#7232]
### What it is?

So we all know just about everything in MemoryMarshal is the type of shit you wouldn't want to whitelist in a million years.

But MemoryMarshal.Cast seems different. [Looking at the docs, it's pretty well protected, all things considered.](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.memorymarshal.cast)

It handles size mismatches for you, and only accept unmanaged types.
I'm not a security expert but my hope is that it's well designed enough that you can't exploit anything with it.

### Why do you need to use this?

It's a long shot but if I'm right and there isn't actually a way to break shit with this function whatsoever, getting this whitelisted would be the timesave of all timesaves.
Especially when you're doing lower level shit like writing to sound streams or building vertex data.
Take this code from a recent project:
```
short[] samples = sampleBuf;
for (int i = 0; i < samples.Length; i++)
{
    samples[i] =
        (short)((data[i * 2 + 0]) |
        (data[i * 2 + 1] << 8));
}
Snd.WriteData(samples);
```
Look at how horrible that is. I've got to go byte by byte in my bigass sound sample array and convert them to shorts.
And I mean yeah you could make the case for a ReadOnlySpan<byte> overload to WriteData, but I've come across this issue again and again, when using Mesh.CreateVertexBuffer/CreateIndexBuffer, GpuBuffer shit, etc. It's everywhere.
I asked in the discord abt this earlier and within 5min I had someone telling me they've had to do the exact same shit in the past.
I can't help but look over the fence and imagine how cool it'd be to just do `Snd.WriteData(MemoryMarshal.Cast<byte, short>(data))` and be done with it, I think the JIT compiler can even optimize these out. Like, getting this whitelisted would make my Christmas.

**Comment by lolleko** [comment]
You are right dotnet runtime catches most out of bounds exploits when using memorymarshal.cast. At least I couldn't figure out how to get around it, but there are most likely smarter people out there that can. So I would rather not whitelist that.

What we may able to do instead is to provide a generic overload to Snd.WriteData and similar functions, that does the marshalling for you.
